### PR TITLE
support rubyzip version 3.0

### PR DESCRIPTION
**Pull request by taichi-ishitani**
Hi @64kramsystem ,

The rubyzip gem is now being developed and its version is `3.0.0`.
https://github.com/rubyzip/rubyzip/blob/master/lib/zip/version.rb

This PR is to fix errors happened when rubyzip v3 is used.

### CODE REVIEW

File: lib/spreadbase/codecs/open_document_12.rb [hunk 1]
```diff
@@ -106,6 +106,15 @@ def decode_content_xml(content_xml_data, options={})
 
       private
 
+      def read_content_xml(zip_buffer)
+        io = StringIO.new(zip_buffer)
+        if Zip::VERSION >= '3.0.0'
```
Comment (64kramsystem) [inline lib/spreadbase/codecs/open_document_12.rb]: This is a semantic mistake :smile::

```
3.0.2 :001 > '10.0.0' > '3.0.0'
 => false 
```

In these cases, one can:

- either wrap both values in `Gem::Version.new`
- or use `Gem.loaded_specs['rubyzip'].version`, and wrap the second value in `Gem::Version.new`

**Review by 64kramsystem — CHANGES_REQUESTED** [review]
There's a semantic mistake :smile:

**Comment by 64kramsystem** [comment]
Hi there! So,

From a user perspective, is there a specific advantage in using Rubyzip 3.0 rather than the current version?

Since the `Gemfile` change is a hack, and Rubyzip 3.0 hasn't even been released, I think the (Spreadbase) users should install the `rubyzip` pre-release version on their own. The change:

```diff
diff --git i/spreadbase.gemspec w/spreadbase.gemspec
index 456d3a6..08df33a 100644
--- i/spreadbase.gemspec
+++ w/spreadbase.gemspec
@@ -17,7 +17,7 @@ Gem::Specification.new do |s|
   s.description = %q{Library for reading/writing OpenOffice Calc documents.}
   s.license     = "GPL-3.0"
 
-  s.add_runtime_dependency     "rubyzip", "~>2.3.0"
+  s.add_runtime_dependency     "rubyzip", ">=2.3.0"
   s.add_development_dependency "rspec",   "~>3.9.0"
 
   s.add_development_dependency "rake",   "~>13.0"
```

could be left in the PR (so that if one wants, they can use Rubyzip 3), then, once v3 is officially released, it can be changed to be `~>3.0.0`. The other changes would also apply (with the exception of `spreadbase.gemspec`).

File: lib/spreadbase/codecs/open_document_12.rb (hunk 1 above)
Comment (taichi-ishitani) [reply lib/spreadbase/codecs/open_document_12.rb]: Thank you for your reviewing!
I've changed the version check logic according to your advice.

**Comment by taichi-ishitani** [comment]
> s.add_runtime_dependency     "rubyzip", ">=2.3.0"

This is for CI purpose.
Bundler cannot resolve version dependency when version constraint is `>~2.3.0` and rubyzip gem is got from github repository.
https://github.com/taichi-ishitani/spreadbase/runs/4646119282?check_suite_focus=true#step:3:32
Therefore I've changed version constraint from `~>` to `>=`.
Do you have any idea to resolve this issue?

**Comment by taichi-ishitani** [comment]
> Do you have any idea to resolve this issue?

https://github.com/taichi-ishitani/spreadbase/blob/84954135396c634f1ce779028855ee657a587230/Gemfile#L10

I found a workaround for this issue but we need to hack Bunder's cache.

**Comment by 64kramsystem** [comment]
> > s.add_runtime_dependency     "rubyzip", ">=2.3.0"
> 
> This is for CI purpose. Bundler cannot resolve version dependency when version constraint is `>~2.3.0` and rubyzip gem is got from github repository. https://github.com/taichi-ishitani/spreadbase/runs/4646119282?check_suite_focus=true#step:3:32 Therefore I've changed version constraint from `~>` to `>=`. Do you have any idea to resolve this issue?

Hello!

So, since this is a significant hack (for a couple of reasons, including using a yet unreleased gem), first, it's important to understand:

- what's the advantage, from a (Spreadbase) user perspective, to use Rubyzip 3.0, rather than 2.3? did you find any spreadsheet that breaks when using 2.3?
- how are you actually installing Rubyzip 3.0 and Spreadbase, and also, make the latter use the former? it can't be done via `gem install spreadbase`, so I take that you're packaging the gem from your fork, and installing it via the `spreadbase-*.gem` file?

Now, regardless of this question, I can accept without any problem a PR (either a new one, or a reduced version of this), which contains just the internal support for Rubyzip 3.0, in other words, this:

```diff
diff --git i/lib/spreadbase/codecs/open_document_12.rb w/lib/spreadbase/codecs/open_document_12.rb
index 5a58842..6a1b94b 100644
--- i/lib/spreadbase/codecs/open_document_12.rb
+++ w/lib/spreadbase/codecs/open_document_12.rb
@@ -59,8 +59,7 @@ def encode_to_archive(el_document, options={})
       # _returns_ the SpreadBase::Document instance.
       #
       def decode_archive(zip_buffer, options={})
-        io = StringIO.new(zip_buffer)
-        content_xml_data = Zip::File.new(io, false, true).read('content.xml')
+        content_xml_data = read_content_xml(zip_buffer)
 
         decode_content_xml(content_xml_data, options)
       end
@@ -106,6 +105,19 @@ def decode_content_xml(content_xml_data, options={})
 
       private
 
+      def read_content_xml(zip_buffer)
+        io = StringIO.new(zip_buffer)
+        if use_rubyzip_3?
+          Zip::File.new(io, buffer: true).read('content.xml')
+        else
+          Zip::File.new(io, false, true).read('content.xml')
+        end
+      end
+
+      def use_rubyzip_3?
+        Gem.loaded_specs['rubyzip'].version >= Gem::Version.create('3.0.0')
+      end
+
       def pretty_xml(document)
         buffer = ""
 
diff --git i/spreadbase.gemspec w/spreadbase.gemspec
index 456d3a6..08df33a 100644
--- i/spreadbase.gemspec
+++ w/spreadbase.gemspec
@@ -17,7 +17,7 @@ Gem::Specification.new do |s|
   s.description = %q{Library for reading/writing OpenOffice Calc documents.}
   s.license     = "GPL-3.0"
 
-  s.add_runtime_dependency     "rubyzip", "~>2.3.0"
+  s.add_runtime_dependency     "rubyzip", ">=2.3.0"
   s.add_development_dependency "rspec",   "~>3.9.0"
 
   s.add_development_dependency "rake",   "~>13.0"
```

This diff allows Spreadbase to support Rubyzip 3.0 - although it doesn't automatically install it.

After this will be merged, we can discuss strategies to install Rubyzip 3.0 and Spreadbase use it.

File: lib/spreadbase/codecs/open_document_12.rb [hunk 2]
```diff
@@ -1,4 +1,5 @@
 require 'zip'
+require 'zip/version'
```
Comment (64kramsystem) [inline lib/spreadbase/codecs/open_document_12.rb]: With the new check, this is not required anymore :slightly_smiling_face:

Comment (taichi-ishitani) [reply lib/spreadbase/codecs/open_document_12.rb]: I've removed it

**Comment by taichi-ishitani** [comment]
> what's the advantage, from a (Spreadbase) user perspective, to use Rubyzip 3.0, rather than 2.3? did you find any spreadsheet that breaks when using 2.3?

Currently, I think there are no advantages to use rubyzip 3.0.

> how are you actually installing Rubyzip 3.0 and Spreadbase, and also, make the latter use the former? it can't be done via gem install spreadbase, so I take that you're packaging the gem from your fork, and installing it via the spreadbase-*.gem file?

I use rubyzip 3.0 for CI only. Therefore, I install it by using bundler and Gemfile.

**Comment by taichi-ishitani** [comment]
I'm integrating your gem with my project and testing with `--enable-frozen-string-literal` option.
However, rubyzip 2.3 does not support the frozen string literal feature completely so I need to use rubyzip 3.0 for my CI.
Rubyzip 3.0 has changes for public APIs and I open this PR to fix errors caused by these public API changes.

**Comment by 64kramsystem** [comment]
(edited to use pessimistic versioning in the spreadbase gem versioning)

Hello!

I've created a PR as previously suggested in the [comment above](https://github.com/64kramsystem/spreadbase/pull/26#issuecomment-1002164776) and released the gem. Now you can use Rubyzip 3.0 in external projects; just structure the project Gemfile like this:

```rb
source 'https://rubygems.org'

gem 'rubyzip', git: 'https://github.com/rubyzip/rubyzip', branch: 'master'
gem 'spreadbase', '~>0.4.0'
```

This strategy doesn't require any Gemfile hack or so.

The change is essentially an extract of this PR, but the logic is so simple that there arern't any significant alternative ways of writing it. But anyway, if you'd like an official mention or commit in the repo, let me know, and I'll do it :smile:

**Comment by taichi-ishitani** [comment]
Hi @64kramsystem ,
Thank you for creating the PR and releasing `0.4.0` version!

**Pull request closed without merging**